internal/graph/cursors.go

```diff
@@ -0,0 +1,267 @@
+
+// withSingletonInCursor executes the given handler if and only if the singleton name is not found
+// at the head of the cursor. If it is found, the handler is skipped. The next handler is always
+// invoked with updated cursor information indicating that the handler's work has been completed.
+func withSingletonInCursor(ci cursorInformation, name string, handler cursorHandler, next cursorHandler) error {
+	skipHandler, err := ci.hasPrefix(name)
+	if err != nil {
+		return err
+	}
+
+	// If the name was found at the head of the cursor, then the handler has already been run and
+	// should be skipped.
+	if !skipHandler {
+		err := handler(ci.mustWithOutgoingSection(name))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Run the next handler.
+	return next(ci.mustWithOutgoingSection(name))
+}
+
+// withIterableInCursor executes the given handler for each item in the items list, skipping any
+// items marked as completed at the head of the cursor and injecting a cursor representing the current
+// item.
+//
+// For example, if items contains 3 items, and the cursor returned was within the handler for item
+// index #1, then item index #0 will be skipped on subsequent invocation.
+func withIterableInCursor[T any](
+	ci cursorInformation,
+	name string,
+	items []T,
+	handler func(ci cursorInformation, item T) error,
+) error {
+	// Check the index for the section in the cursor. If found, we skip any items before that index.
+	afterIndex, err := ci.integerSectionValue(name)
+	if err != nil {
+		return err
+	}
+
+	for index, item := range items {
+		if index < afterIndex {
+			continue
+		}
+
+		// Invoke the handler with the current item's index in the outgoing cursor, indicating that
+		// subsequent invocations should jump right to this item.
+		err := handler(ci.mustWithOutgoingSection(name, strconv.Itoa(index)), item)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// withQueryInCursor executes the given handler until it returns an empty "next" datastore cursor,
+// starting at the datastore cursor found in the cursor information (if any).
+func withQueryInCursor(
+	ci cursorInformation,
+	name string,
+	handler func(queryCursor options.Cursor, ci cursorInformation) (options.Cursor, error),
+) error {
+	// Retrieve the *datastore* cursor, if one is found at the head of the incoming cursor.
+	var datastoreCursor options.Cursor
+	datastoreCursorString, err := ci.sectionValue(name)
+	if err != nil {
+		return err
+	}
+
+	if datastoreCursorString != "" {
+		datastoreCursor = tuple.MustParse(datastoreCursorString)
+	}
+
+	// Execute the loop, starting at the datastore's cursor (if any), until there is no additional
+	// datastore cursor returned.
+	for {
+		currentCursor := ci.mustWithOutgoingSection(name, tuple.MustString(datastoreCursor))
+		nextDCCursor, err := handler(datastoreCursor, currentCursor)
+		if err != nil {
+			return err
+		}
+		if nextDCCursor == nil {
+			return nil
+		}
+		datastoreCursor = nextDCCursor
+	}
```

Comment:
Can you explain how this loop is honoring the API request-provided limit? This is calling the datastore with a limit of 100, but iterating until the datastore exhausts the results. Nothing seems to check "ok, we've reached the limit, stop and return the cursor wherever we are at".

Reply:
See my explanation above: for datastore loads, we usually need to load *more* intermediate objects in order to find the specified limit of resources.

We do early terminate the loop here by checking the limit via the limit tracker, which will also cancel the context automatically once the limit is reached